### PR TITLE
feat(mapPage): selection logic for map page

### PR DESCRIPTION
**Pull request by firestack**
I split out the autocenterer for this so that the offset for the mapPage can be configured independently of the other maps in skate. 

---

Asana Ticket: https://app.asana.com/0/1203014709808707/1203641439240315/f

### CODE REVIEW

File: assets/tests/testHelpers/leafletMapWidth.ts
Comment (firestack) [inline assets/tests/testHelpers/leafletMapWidth.ts:38]: I couldn't figure out a good default, some tests have been built around 0,0 without knowing it, I think it's best to follow what jsdom is telling us, but I could be convinced otherwise.

Advice for where to put this documentation would be welcome.

File: assets/tests/testHelpers/leafletMapWidth.ts
Comment (firestack) [inline assets/tests/testHelpers/leafletMapWidth.ts:44]: not super happy that this was a side-effect of importing, but it made tests a lot more stable IIRC. Should probably revisit?

File: assets/tests/testHelpers/wrappers.tsx
Comment (firestack) [inline assets/tests/testHelpers/wrappers.tsx:24]: Still iffy on this but tests that interact with the global state needed this to reduce boilerplate and complexity in tests.

As a behavioral testing tool, it's very useful.

File: assets/src/components/mapPage.tsx
Comment (lemald) [inline assets/src/components/mapPage.tsx:136]: **suggestion:** I think you can remove the `=== true` here because `vpcEnabled` was created from the `Boolean` constructor already.

Comment (firestack) [reply assets/src/components/mapPage.tsx:136]: I know I _can_ but I really prefer the explicitness of how it reads. I'm happy to remove it, but I do think this is better for future readers of the code.

File: assets/src/components/mapPage.tsx
Comment (lemald) [inline assets/src/components/mapPage.tsx:177]: **question:** is this fragment a placeholder for something? Right now I don't think it actually changes anything about the DOM if `selectedVehicle` is present and also a `Vehicle`.

Comment (firestack) [reply assets/src/components/mapPage.tsx:177]: yeah that's where the autocenterer is going

File: assets/src/components/mapPage.tsx
Comment (lemald) [reply assets/src/components/mapPage.tsx:136]: I guess the way I think of it is that the name `vpcEnabled` already connotes that it's a boolean, although sadly JS / TS lacks an actual convention for naming boolean values like putting a `?` at the end in Elixir so it's not explicit as it could be. Since it's purely subjective I'm fine with leaving it the way you prefer.

File: assets/tests/components/map.test.tsx
Comment (lemald) [reply assets/tests/components/map.test.tsx:347]: If you have time and feel like it, I'd be happy to look at a PR for that refactor! Otherwise I agree that these assertions aren't _great_ but I don't think they're dealbreakers.

File: assets/tests/testHelpers/leafletMapWidth.ts
Comment (lemald) [reply assets/tests/testHelpers/leafletMapWidth.ts:38]: I like the approach of going with what jsdom tells us. Screen dimensions are one of those things I've never really considered very much when writing tests since we're usually so focused on the raw DOM structure, so thank you for doing some work to make it handled somewhat more correctly!

File: assets/src/components/map.tsx
Comment (KaylaBrady) [inline assets/src/components/map.tsx:239]: Appreciate this name change!

File: assets/src/components/map.tsx
Comment (KaylaBrady) [inline assets/src/components/map.tsx:249]: question (non-blocking): I realize this mostly from the existing logic so doesn't really need to change, but is there a case when these would not be equal since `currentLatLngs` is initialized to `positions`?

File: assets/src/components/map.tsx
Comment (KaylaBrady) [inline assets/src/components/map.tsx:266]: nitpick (non-blocking): The name `isUpdating` was a bit confusing to me at first with the combination of the name `InteractiveFollowState` - I didn't quite get what it is that is being updated, in combination with it being set to `isFollowing` on line 280 below.

File: assets/src/components/map.tsx
Comment (KaylaBrady) [inline assets/src/components/map.tsx:287]: question: should this be `isFollowing` or `isUpdating` for consistency?

File: assets/src/components/map.tsx
Comment (KaylaBrady) [inline assets/src/components/map.tsx:337]: question (non-blocking): Why `useCallback` here? I'm guessing it prevents re-renders, but I'm not too familiar with this hook

File: assets/src/components/mapPage.tsx
Comment (KaylaBrady) [inline assets/src/components/mapPage.tsx:164]: question: would these vehicles not be primary, similar to the non-selected vehicles in the VPP?

File: assets/src/components/mapPage.tsx
Comment (KaylaBrady) [inline assets/src/components/mapPage.tsx:280]: question (non-blocking): Are these empty tags needed?

**Review by KaylaBrady — COMMENTED** [review]
This is looking good! Still wrapping my head around some of the following/selecting logic which may take a few more reads on my part, but I think some of the discussion from these questions will help me get there!

File: assets/src/components/map.tsx
Comment (lemald) [inline assets/src/components/map.tsx:239]: **praise:** I like the updated terminology better than "autocenter."

File: assets/src/components/mapPage.tsx
Comment (firestack) [reply assets/src/components/mapPage.tsx:280]: nope! I actually cleaned this up when implementing the VPC ghost card https://github.com/mbta/skate/pull/1895
I was considering cherry picking that commit into this PR

File: assets/src/components/mapPage.tsx
Comment (firestack) [reply assets/src/components/mapPage.tsx:111]: when `useVehicleForId` changes which topic it's subscribed to, it initially returns `null` which makes the VPC and route shape, route vehicles, and selected vehicle, blip in and out of existance as it rapidly rerenders.

I cache the last successful result before `selectedVehicleId` changes so that we don't blip the UI while loading the next vehicle.
I had this working with a reducer/state, but because what triggers the render is `useVehicleForId` updating, we only need to hold onto the previous vehicle till the next one is available/not null WHILE the `selectedVehicleId` value tells us that we didn't deselect a vehicle by not being `null`.

Maybe naming this `useDiferredVehicleById` would be better?... though this doesn't match what `useDiferredValue` does exactly. (I also tried to get this to work with `useDiferredValue` but that didn't work because `useSelectedVehicleForId` may return `null` multiple times before the next value/vehicle has resolved)

File: assets/src/components/mapPage.tsx
Comment (firestack) [reply assets/src/components/mapPage.tsx:164]: I dunno! I can check with design on that, from the mockups I'd seen it seemed better to go with the primary vehicle style.

File: assets/src/components/mapPage.tsx
Comment (firestack) [reply assets/src/components/mapPage.tsx:202]: Because it was easier to implement for MVP heh.
The logic to fit a bounds would be similar but need a bit more thinking/work, so I left that to the future.

I wasn't sure how to allow the follower to change the type of it's `positions` prop based on the `onUpdate` function at the time of implementaiton, but that could probably be done?(unsure how long it'd take)
But then the component should be able to take a `LatLng | null` instead of `LatLng[]` if the component could change the type of the `positions` prop based on the second argument type of the `onUpdate` prop.

File: assets/src/components/map.tsx
Comment (firestack) [reply assets/src/components/map.tsx:337]: I did it to prevent rerenders, but I don't actually remember if it was needed. It seemed like it was causing issues but it might have just been something I did while debugging and didn't reevaluate.

File: assets/src/components/map.tsx
Comment (firestack) [reply assets/src/components/map.tsx:287]: Yep! Thank you!

File: assets/src/components/map.tsx
Comment (firestack) [reply assets/src/components/map.tsx:249]: I didn't really question this becuase it was the existing logic. I think this is doing a value check rather than using the `Object.is` check that `useEffect` does to keep them updated only when the actual values of `positions` change?

Otherwise, because the `Object.is` function would return `null` for two different instances of arrays of the same data, I think the `useEffect` would fire constantly if it was dependant on `positions`? 
This should probably be checked.

File: assets/src/components/map.tsx
Comment (firestack) [reply assets/src/components/map.tsx:266]: thank you!

Comment (firestack) [reply assets/src/components/map.tsx]: let me know what you think of 7be135c4013b264282c75fe8618aca76dff3e0d7 ?

File: assets/src/components/map.tsx
Comment (firestack) [reply assets/src/components/map.tsx]: https://github.com/mbta/skate/pull/1891#discussion_r1085875987

File: assets/src/components/map.tsx
Comment (lemald) [reply assets/src/components/map.tsx:249]: @firestack yes, that's all correct as to why it works this way currently. @KaylaBrady if I'm understanding your question correctly, basically on the first render they will be equal, but on subsequent renders the `positions` prop will be updated but `currentLatLngs` will have the old value. Then, the state is only updated if there's an actual change in the contents.

File: assets/src/components/mapPage.tsx
Comment (lemald) [reply assets/src/components/mapPage.tsx:111]: Just spitballing here, how about `useMostRecentVehicleById`?

Comment (firestack) [reply assets/src/components/mapPage.tsx:111]: @lemald 👍🏻👍🏻

**Review by lemald — DISMISSED** [review]
Okay, I've finally done a start-to-finish read over and I think this looks good, with the only issues I'm aware of being those already in flight. 👍

File: assets/src/components/mapPage.tsx
Comment (firestack) [reply assets/src/components/mapPage.tsx:202]: I just had it come to me that I was wrong about implementation difficulty, I'll make a patch for review!

File: assets/src/components/map.tsx
Comment (KaylaBrady) [reply assets/src/components/map.tsx]: Thank you - `isAnimatingFollowUpdate` is so clear!!

File: assets/src/components/mapPage.tsx
Comment (KaylaBrady) [reply assets/src/components/mapPage.tsx:164]: Ah yeah I see what you mean! A separate question in this same area - is there an advantage to creating the markers in the `MapPage` component rather than delegating that to the existing `BaseMap` functionality using the `vehicles` and `selectedVehicleId` props?

File: assets/src/components/mapPage.tsx
Comment (KaylaBrady) [reply assets/src/components/mapPage.tsx:202]: Ohh maybe I missed something about this the first time around - is points is only ever a list of size one? I had it in my head that it could be a list including all the visible vehicles

File: assets/src/components/mapPage.tsx
Comment (firestack) [reply assets/src/components/mapPage.tsx:164]: The biggest issue was the autocenterer used all "primary" vehicles by default, and not wanting to add more props/functionality to the base component to support the map page configuration, and "idiomatic" react seemed to be pushing me in the direction of composition over "inheritance" (adding more props..?) to make the map easier to configure for all the different use cases.

I've actually been thinking that we might want to slim down the props that `BaseMap` takes and provide more functionality via composition instead (and preconfigued wrapper components, like the MiniMap). As well as maybe grouping our markers into `pane`s so that it's easier to look through in the react dev tools.

File: assets/src/components/mapPage.tsx
Comment (firestack) [reply assets/src/components/mapPage.tsx:202]: Yep! https://github.com/mbta/skate/pull/1891/files#diff-a6a64d5b0e8199bb4a7ea39da345d5247083943b92fd078e359cb4d1d9ab60ffR259

File: assets/src/components/map.tsx
Comment (firestack) [reply assets/src/components/map.tsx:337]: So this `useCallback` doesn't actually seem to do anything? so I'll remove it!

File: assets/src/components/mapPage.tsx
Comment (firestack) [reply assets/src/components/mapPage.tsx:164]: Design said yes to the non selected vehicles not being primary 👍🏻

File: assets/src/components/mapPage.tsx
Comment (firestack) [reply assets/src/components/mapPage.tsx]: @mbta/skate-developers please take a look at [d90c0e53](https://github.com/mbta/skate/pull/1891/commits/d90c0e53016f508ce9a39a7d24ebc82fff14a66a) 🙏🏻

File: assets/src/components/mapPage.tsx
Comment (firestack) [reply assets/src/components/mapPage.tsx]: I think I covered everything in this thread, so I'm gonna resolve this, but I'll pay attention to this thread if anyone wants to point anything else out!

File: assets/src/components/map.tsx
Comment (firestack) [reply assets/src/components/map.tsx]: [`efb9a92` (#1891)](https://github.com/mbta/skate/pull/1891/commits/efb9a92e7648a720e61c66030b73133176ffe0d2)

File: assets/src/components/mapPage.tsx
Comment (firestack) [reply assets/src/components/mapPage.tsx]: Cherry picked into [`c536580` (#1891)](https://github.com/mbta/skate/pull/1891/commits/c53658007572b47cc69963b14d988ae50cde0875)

**Comment by firestack** [comment]
The only open question for code review I'm aware of is https://github.com/mbta/skate/pull/1891#discussion_r1086864229

Let me know if there's anything else, otherwise once design & product approve I'll merge 👍🏻

File: assets/src/components/mapPage.tsx
Comment (KaylaBrady) [reply assets/src/components/mapPage.tsx]: This looks good to me! I'm still working on understanding the benefits of `useRef` over `useState` in this case, but maybe I just need to revisit the `useRef` docs & don't want to block merging over it!

File: assets/src/components/mapPage.tsx
Comment (lemald) [inline assets/src/components/mapPage.tsx]: **comment (non-blocking):** I realize the intent of splitting this into two separate assignments is to avoid calling hooks conditionally, but usually we use the convention from Elixir (and maybe other languages?) where a variable name prefixed with an underscore isn't used later on, so this feels _slightly_ at odds with that convention, although its one use is in the came `const` statement... so I'm conflicted about how I feel about it.

File: assets/src/components/mapPage.tsx
Comment (lemald) [reply assets/src/components/mapPage.tsx]: One nitpick-y comment about variable names but otherwise I like it!

File: assets/src/components/mapPage.tsx
Comment (firestack) [reply assets/src/components/mapPage.tsx]: maybe we can setup a lint rule so that we get warned when variables with a leading `_` are used?

File: assets/src/components/mapPage.tsx
Comment (firestack) [reply assets/src/components/mapPage.tsx]: > benefits of useRef over useState in this case

Here's my understanding on what would happen if we used `useState` instead of `useRef`. 

When the `useVehicleForId` hook updates asynchronously, the state update causes `<MapDisplay/>` to rerender.

On that next render, `MapDisplay` can save the result of `useVehicleForId` using either `useState` or `useRef`.

When `useVehicleForId` updates, we need to store the value in case the next render returns `null` (which should only happen when `selectedVehicleId` changes and is not `null`). [^1]


__If `useVehicleForId` is cached in a `useState`, then updating the cache will cause a rerender, which would mean for every `useVehicleForId` update we would render the `<MapDisplay/>` twice.__



[^1]: Actually, maybe we can catch the last return value of `useVehicleForId` when `selectedVehicleId` changes? But unless the lambda in a `useEffect(..., [selectedVehicleId])` can capture the `VehicleOrGhost` before/while it changes to `null`, I still think we'd need a ref or state to track `lastSelectedVeihcleId`?

Comment (firestack) [reply assets/src/components/mapPage.tsx]: I'm still learning react so maybe there's something I missed! but 'TLDR;' I picked `useRef` because I didn't want to cause a rerender when updating it.

**Comment by firestack** [comment]
I'm going to put this into draft until I confirm about final Acceptance Criteria, will reopen when ready for review.

**Comment by firestack** [comment]
After design/product review I'm a little unsure of what the AC is around test group users, as it's a little unclear if these changes should also be done for users not in the test group. This PR should only affect users with access to the MapPage, and I've opened https://github.com/mbta/skate/pull/1897 in the case that we do want some of these changes pushed to users.

I'd like to be able merge this to unblock any PR's waiting on it, but we need final approval from product first, so I've split this out into two PR's.

**Review by lemald — APPROVED** [review]

**Pull request merged**
